index.js

```diff
@@ -10,50 +12,96 @@ var getCallerFile = require('get-caller-file');
 var loadTSConfig = require('./lib/load-ts-config');
 var findTSConfig = require('./lib/find-ts-config');
 
+var Promise = require('rsvp').Promise;
+
 module.exports = TypeScript;
+
 function TypeScript(inputTree, _options) {
+  if (!(this instanceof TypeScript)) {
+    return new TypeScript(inputTree, _options);
+  }
+
   var options = _options || {};
-  options.tsconfig = options.tsconfig || findTSConfig(getCallerFile(2));
+  var config = options.tsconfig || findTSConfig(getCallerFile(2));
 
-  if (!(this instanceof TypeScript)) {
-    return new TypeScript(inputTree, options);
+  if (typeof config === 'string') {
+    this.configFileName = config;
+    config = readConfigFile(config);
+  } else {
+    this.configFileName = undefined;
   }
+  this.config = config || {};
+  this.program = undefined;
 
-  Filter.call(this, inputTree, {
-    persist: true,
-    extensions: ['ts','d.ts'],
-    targetExtension: 'js',
+  Plugin.call(this, [ inputTree ], {
     name: 'broccoli-typescript-compiler',
+    persistentOutput: true,
     annotation: options.annotation
   });
-
-  this.options = loadTSConfig(options.tsconfig);
 }
 
-TypeScript.prototype = Object.create(Filter.prototype);
+TypeScript.prototype = Object.create(Plugin.prototype);
 TypeScript.prototype.constructor = TypeScript;
 
 TypeScript.prototype.baseDir = function() {
   return __dirname;
 };
 
+TypeScript.prototype.build = function build() {
+  var inputPath = this.inputPaths[0];
+  var config = ts.parseJsonConfigFileContent(this.config, ts.sys, inputPath, undefined, this.configFileName);
+  this.compile(config);
+  return Promise.resolve();
+}
+
+TypeScript.prototype.compile = function (config) {
+
+  var compilerHost = ts.createCompilerHost(config.options);
+
+  var inputPath = this.inputPaths[0];
+  var outputPath = this.outputPath;
+
+  var hostWriteFile = compilerHost.writeFile;
+  compilerHost.writeFile = function (fileName, data, writeByteOrderMark, onError) {
+    var outFileName = outputPath + fileName.substr(inputPath.length);
+    hostWriteFile(outFileName, data, writeByteOrderMark, onError);
```

Comment:
I wonder if the `host` object can provide some protocol for detecting if something is stale ornot.